src/application/AnimalSubmissionsService.ts

```diff
@@ -93,12 +95,22 @@ export class AnimalSubmissionsService {
                 throw new ApiError('Unauthorized', 401, 'User and volunteer can only get own submissions');
             }
         }
+
+        const isFirstPage = paginationParams?.page == 1 ? true : false;
+
+        const SKIP =
+            paginationParams?.perPage && paginationParams?.page ? paginationParams.perPage * paginationParams.page : 0;
+
+        const LIMIT = paginationParams?.perPage ? paginationParams.perPage : undefined;
+
```

Comment:
Można by sprawdzać, czy użytkownik podał `paginationParams`, tak aby nie rzucał błędami typu `Cannot read property perPage of undefined` (chodzi o znaki zapytania przy obiekcie `paginationParams`). Jeśli nie to tworzyć obiekt z zdefiniowanymi parametrami domyślnymi lub coś innego 😅.

Reply:
@Haivex 
Chodzi mi o to, że jeżeli przyjmujesz w parametrze funkcji `paginationParams?: PaginationParams`, to jest równoznaczne zapisowi `paginationParams: PaginationParams | undefined`. Jeśli nie podasz wartości, wtedy `paginationParams` ma wartość `undefined`. Z tego powodu próbując odwołać się do parametru `paginationParams.perPage` musisz dopisać znak zapytania w taki sposób `paginationParams?.perPage`, co wiąże się z rzucaniem błędów typu `Cannot read property of undefined`.
Aby temu zapobiec, można sprawdzić, czy użytkownik podał obiekt `paginationParams`, a dopiero później wykonać dalszą część kodu. Można to rozwiązać w ten sposób
```ts
public async getAllAnimalSubmissions(
        queryParams: GetAllAnimalSubmissionsParams,
        currentUser: IUserInfo,
        paginationParams?: PaginationParams,
    ): Promise<FormAnimalSubmission[]> {
        if (currentUser.role == UserType.NORMAL || currentUser.role == UserType.VOLUNTEER) {
            const submission = await this.animalSubmissionRepository
                .createQueryBuilder('submission')
                .leftJoinAndSelect('submission.applicant', 'applicant')
                .where('applicant.id = :id', { id: currentUser.id })
                .getOne();

            if (submission?.applicant.id != currentUser.id) {
                throw new ApiError('Unauthorized', 401, 'User and volunteer can only get own submissions');
            }
        }
        if (!paginationParams) {
            const submissions = await new OptionalWhereSelectQueryBuilder(
                this.animalSubmissionRepository
                    .createQueryBuilder('submission')
                    .leftJoinAndSelect('submission.animal', 'animal')
                    .leftJoinAndSelect('submission.applicant', 'applicant')
                    .leftJoinAndSelect('submission.reviewer', 'reviewer'),
            )
                .optAndWhere('submission.status = ', queryParams.status)
                .optAndWhere('submission.submissionDate = ', queryParams.submissionDate)
                .optAndWhere('animal.name = ', queryParams.animalName)
                .optAndWhere('animal.specie = ', queryParams.specie)
                .optAndWhere('applicant.name = ', queryParams.userName)
                .optAndWhere('reviewer.name = ', queryParams.reviewerName)
                .selectQueryBuilder.getMany();

            if (submissions.length === 0) throw new ApiError('Not Found', 404, 'Submissions not found');
            return submissions;
        }

        const isFirstPage = paginationParams.page == 1 ? true : false;
        const SKIP =
            paginationParams.perPage && paginationParams.page ? paginationParams.perPage * paginationParams.page : 0;
        const LIMIT = paginationParams.perPage ? paginationParams.perPage : undefined;

        const submissions = await new OptionalWhereSelectQueryBuilder(
            this.animalSubmissionRepository
                .createQueryBuilder('submission')
                .leftJoinAndSelect('submission.animal', 'animal')
                .leftJoinAndSelect('submission.applicant', 'applicant')
                .leftJoinAndSelect('submission.reviewer', 'reviewer')
                .skip(isFirstPage ? 0 : SKIP)
                .limit(LIMIT),
        )
            .optAndWhere('submission.status = ', queryParams.status)
            .optAndWhere('submission.submissionDate = ', queryParams.submissionDate)
            .optAndWhere('animal.name = ', queryParams.animalName)
            .optAndWhere('animal.specie = ', queryParams.specie)
            .optAndWhere('applicant.name = ', queryParams.userName)
            .optAndWhere('reviewer.name = ', queryParams.reviewerName)
            .selectQueryBuilder.getMany();

        if (submissions.length === 0) throw new ApiError('Not Found', 404, 'Submissions not found');
        return submissions;
    }
```
Teraz jeżeli użytkownik nie podał obiektu `paginationParams`, nie będzie żadnej paginacji, chociaż to zachowanie można zmienić.

Reply:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining
W zmiennych SKIP i LIMIT sprawdzam czy perPage i page są undefined jeśli są to metody skip i limit wykonują się z wartością undefined bo mogą, co oznacza po prostu że nie ma żadnej paginacji. To co ty zaproponowałeś to duplikacja kodu.

Reply:
Sprawdzałem, z podawaniem parametru i bez :)

Reply:
Jest brzydki kod z znakiem zapytania (złe praktyki, bo próbujesz się odwołać do właściwości obiektu, który ma wartość `undefined`), jak chcesz żeby zawsze była paginacja to można podać wartości domyślne. 
Jeśli nie chcesz żadnej paginacji w domyśle paginacji możesz zrobić coś takiego jeszcze:
```ts
public async getAllAnimalSubmissions(
        queryParams: GetAllAnimalSubmissionsParams,
        currentUser: IUserInfo,
        paginationParams?: PaginationParams,
    ): Promise<FormAnimalSubmission[]> {
        if (currentUser.role == UserType.NORMAL || currentUser.role == UserType.VOLUNTEER) {
            const submission = await this.animalSubmissionRepository
                .createQueryBuilder('submission')
                .leftJoinAndSelect('submission.applicant', 'applicant')
                .where('applicant.id = :id', { id: currentUser.id })
                .getOne();

            if (submission?.applicant.id != currentUser.id) {
                throw new ApiError('Unauthorized', 401, 'User and volunteer can only get own submissions');
            }
        }
        let isFirstPage;
        let skip;
        let limit;
        if (paginationParams) {
            isFirstPage = paginationParams.page == 1 ? true : false;
            skip =
                paginationParams.perPage && paginationParams.page
                    ? paginationParams.perPage * paginationParams.page
                    : 0;
            limit = paginationParams.perPage ? paginationParams.perPage : undefined;
        }

        const submissions = await new OptionalWhereSelectQueryBuilder(
            this.animalSubmissionRepository
                .createQueryBuilder('submission')
                .leftJoinAndSelect('submission.animal', 'animal')
                .leftJoinAndSelect('submission.applicant', 'applicant')
                .leftJoinAndSelect('submission.reviewer', 'reviewer')
                .skip(isFirstPage ? 0 : skip)
                .limit(limit),
        )
            .optAndWhere('submission.status = ', queryParams.status)
            .optAndWhere('submission.submissionDate = ', queryParams.submissionDate)
            .optAndWhere('animal.name = ', queryParams.animalName)
            .optAndWhere('animal.specie = ', queryParams.specie)
            .optAndWhere('applicant.name = ', queryParams.userName)
            .optAndWhere('reviewer.name = ', queryParams.reviewerName)
            .selectQueryBuilder.getMany();

        if (submissions.length === 0) throw new ApiError('Not Found', 404, 'Submissions not found');
        return submissions;
    }
```
Dodałem 3 linijki i zabezpieczyłem się przed odczytaniem propsów z undefined

Reply:
Ogarnąłem :)